Fix VersionWithType Crossgen2 check for parameterized types

trylek opened the pull request:
As Egor Bogatov discovered in

https://github.com/dotnet/runtime/pull/90441

Crossgen2 is unable to compile a method taking RuntimeTypeMethod[] parameter because Crossgen2 decides that the array doesn't version with the module being compiled (System.Private.CoreLib in this case). This change modifies the relevant logic so that, for parameterized types, we decide on versioning based on the parameter type.

Thanks

Tomas

/cc @dotnet/crossgen-contrib

trylek committed:
Fix VersionWithType Crossgen2 check for parameterized types

As Egor Bogatov discovered in

https://github.com/dotnet/runtime/pull/90441

Crossgen2 is unable to compile a method taking RuntimeTypeMethod[]
parameter because Crossgen2 decides that the array doesn't version
with the module being compiled (System.Private.CoreLib in this
case). This change modifies the relevant logic so that, for
parameterized types, we decide on versioning based on the
parameter type.

Thanks

Tomas

trylek commented:
/azp run runtime-coreclr outerloop

trylek commented:
/azp run runtime-coreclr crossgen2 outerloop

trylek commented:
Just as a heads-up the crossgen2 outerloop failures are expected, the pipeline hasn't been clean for quite a while, I hope to be able to spend some time on it as part of quality weeks.

jkotas commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs:
```diff
@@ -363,8 +363,15 @@ public sealed override bool VersionsWithModule(ModuleDesc module)
         {
-            return typeDesc.GetTypeDefinition() is EcmaType ecmaType &&
-                _versionsWithTypeCache.GetOrAdd(typeDesc, _versionsWithTypeUncached);
+            if (typeDesc.GetTypeDefinition() is EcmaType)
```
It does not look right that we handle parametrized arrays here and generics in ComputeTypeVersionsWithCode. 

Should this method just "return _versionsWithTypeCache.GetOrAdd(typeDesc, _versionsWithTypeUncached)` and the handling of parametrized types to be all in ComputeTypeVersionsWithCode?

jkotas commented:
I think you should be able to delete https://github.com/dotnet/runtime/blob/09af161453ffc7ef8ac9b7175167602f1de32313/src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs#L918-L922 with this fix - it is a local workaround for the problem that this is fixing in more general way.

trylek committed:
Address JanK's PR feedback

Based on Jan Kotas' feedback I have moved the logic for determining
versionability of parameterized types into the method
ComputeTypeVersionsWithCode and I have removed the two specific
checks for IsArray / IsPointer in the instantiation validation
logic that should be superseded by this more general conditional
statement.

In local testing I noticed that three negative variance tests fail
when built with Crossgen2 but that's apparently unrelated to my
change, I see the same three failures in a build off main. I plan
to look into health of the Crossgen2 outerloop pipeline next.

Thanks

Tomas

trylek commented:
/azp run runtime-coreclr outerloop

trylek committed:
Cleanup in ComputeInstantiationTypeVersionsWithCode

After removal of the IsArray and IsPointer checks the method has
become mostly trivial so I have slightly simplified its body.

Thanks

Tomas

jkotas reviewed (approved):
Thank you for cleaning it up!

trylek commented:
The OSX Helix machines are super slow so my retried run won't finish before the 5PM cutoff but the first run only showed one failure in the <code>BinderTracingTest.ResolutionFlow</code> that keeps popping up sporadically and I believe it to be unrelated to my change, merging in to make it to the RC1 cutoff. If the BinderTracingTest turns out to be a new serious issue, it will require addressing one way or the other.
